thrifty-runtime/src/main/java/com/microsoft/thrifty/service/AsyncClientBase.java

```diff
@@ -0,0 +1,267 @@
+    protected void enqueue(MethodCall<?> methodCall) {
+        if (!running.get()) {
+            throw new IllegalStateException("Cannot write to a closed service client");
+        }
+
+        if (!pendingCalls.offer(methodCall)) {
+            // This should never happen with an unbounded queue
+            throw new IllegalStateException("Call queue is full");
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        close(null);
+    }
+
+    private void close(final Throwable error) {
+        if (!running.compareAndSet(true, false)) {
+            return;
+        }
+
+        workerThread.interrupt();
+
+        try {
+            protocol.close();
+        } catch (IOException ignored) {
+            // nope
+        }
+
+        if (!pendingCalls.isEmpty()) {
+            List<MethodCall<?>> incompleteCalls = new ArrayList<>();
+            pendingCalls.drainTo(incompleteCalls);
+            CancellationException e = new CancellationException();
+            for (MethodCall<?> call : incompleteCalls) {
+                try {
+                    fail(call, e);
+                } catch (Exception ignored) {
+                    // nope
+                }
+            }
+        }
+
+        callbackExecutor.execute(new Runnable() {
+            @Override
+            public void run() {
+                if (error != null) {
+                    listener.onError(error);
+                } else {
+                    listener.onTransportClosed();
+                }
+            }
+        });
+
+        try {
+            // Shut down, but let queued tasks finish.
+            // Don't terminate!
+            callbackExecutor.shutdown();
+        } catch (Exception ignored) {
+            // nope
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    private class WorkerThread extends Thread {
+        @Override
+        public void run() {
+            Throwable error = null;
+            while (running.get()) {
+                try {
+                    invokeRequest();
+                } catch (Throwable e) {
+                    error = e;
+                    break;
+                }
+            }
+
+            try {
+                close(error);
+            } catch (Throwable ignored) {
+                // nope
+            }
+        }
+
+        private void invokeRequest() throws IOException, InterruptedException {
+            MethodCall<?> call = pendingCalls.take();
+            if (!running.get()) {
+                if (call != null) {
+                    fail(call, new CancellationException());
+                }
+                return;
+            }
+
+            if (call == null) {
+                return;
+            }
+
+            Object result = null;
+            Exception error = null;
+            try {
+                result = AsyncClientBase.this.invokeRequest(call);
+            //TODO a nicer way would be a struct exception base class and only catch that
```

Comment:
See todo. Do we want something like a `StructException` as base class for the explicitly declared exceptions which are passed to `fail()`?